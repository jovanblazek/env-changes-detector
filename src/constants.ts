export const OUTPUT = {
  HAS_DETECTED_CHANGES: 'env-changes-detected',
  RAW: 'env-changes-raw',
  MARKDOWN: 'env-changes-md'
}

export const MARKDOWN_MESSAGE = {
  NO_CHANGES: 'No env file changes detected.',
  CHANGES_DETECTED: 'Detected changes in env files:',
  FILE_RENAMED: 'File was renamed.',
  FILE_DELETED: 'File was deleted.',
  PARSING_ERROR: 'Could not determine what has changed. Manual check required.'
}
