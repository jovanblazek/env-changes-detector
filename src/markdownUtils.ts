export const getFileDiffHeader = (fileName: string): string =>
  `#### \`${fileName}\`\n`

export const getRenamedFileDiffHeader = (
  pathBefore: string,
  pathAfter: string
): string => `#### \`${pathBefore} -> ${pathAfter}\`\n`

export const getMarkdownDiffWrapper = (diff: string): string =>
  `\`\`\`diff\n${diff}\n\`\`\`\n`
