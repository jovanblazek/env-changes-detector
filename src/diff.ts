import { exec } from 'child_process'
import { promisify } from 'util'
import parseGitDiff, {
  type AnyChunk,
  type AnyFileChange,
  type AnyLineChange
} from 'parse-git-diff'
import { MARKDOWN_MESSAGE } from './constants'
import {
  getFileDiffHeader,
  getMarkdownDiffWrapper,
  getRenamedFileDiffHeader
} from './markdownUtils'

export const getRawDiff = async (
  targetBranch: string,
  filesPattern: string[]
): Promise<string> => {
  const filesPatternString = filesPattern.map(file => `'${file}'`).join(' ')
  const { stdout, stderr } = await promisify(exec)(
    `git diff -w -M100% origin/${targetBranch} -- ${filesPatternString}`
  )
  if (stderr) {
    throw new Error(stderr)
  }
  return stdout
}

const convertLineChangeToMarkdown = (lineChange: AnyLineChange): string => {
  const { type, content } = lineChange
  switch (type) {
    case 'AddedLine':
      return `+ ${content}`
    case 'DeletedLine':
      return `- ${content}`
    default:
      return ''
  }
}

const convertChunkToMarkdown = (chunk: AnyChunk): string[] => {
  const { type } = chunk
  switch (type) {
    case 'Chunk':
    case 'CombinedChunk':
      return chunk.changes.map(convertLineChangeToMarkdown).filter(Boolean) // Remove empty strings
    default:
      return []
  }
}

const getFileDiff = (file: AnyFileChange): string => {
  const { type, chunks } = file
  const fileDiff = chunks.flatMap(convertChunkToMarkdown).join('\n')
  if (!fileDiff.length && (type === 'AddedFile' || type === 'ChangedFile')) {
    throw new Error(`Could not parse diff for file: ${file.path}.`)
  }

  switch (type) {
    case 'AddedFile':
    case 'ChangedFile': {
      return `${getFileDiffHeader(file.path)}${getMarkdownDiffWrapper(
        fileDiff
      )}`
    }
    case 'RenamedFile': {
      return `${getRenamedFileDiffHeader(
        file.pathBefore,
        file.pathAfter
      )}${getMarkdownDiffWrapper(MARKDOWN_MESSAGE.FILE_RENAMED)}`
    }
    case 'DeletedFile': {
      return `${getFileDiffHeader(file.path)}${getMarkdownDiffWrapper(
        MARKDOWN_MESSAGE.FILE_DELETED
      )}`
    }
    default:
      return ''
  }
}

export const getMarkdownDiff = (rawDiff: string): string => {
  const hasDetectedChanges = rawDiff.length > 0

  if (!hasDetectedChanges) {
    return MARKDOWN_MESSAGE.NO_CHANGES
  }

  const { files } = parseGitDiff(rawDiff)
  let markdownDiff = ''

  for (const file of files) {
    try {
      const fileDiff = getFileDiff(file)
      markdownDiff += `${fileDiff}\n`
    } catch {
      markdownDiff += `${getFileDiffHeader(
        file.type === 'RenamedFile'
          ? getRenamedFileDiffHeader(file.pathBefore, file.pathAfter)
          : file.path
      )}${MARKDOWN_MESSAGE.PARSING_ERROR}\n`
    }
  }

  return `## ${MARKDOWN_MESSAGE.CHANGES_DETECTED}\n${markdownDiff}`
}
