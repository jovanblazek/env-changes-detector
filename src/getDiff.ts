import { exec } from 'child_process'
import { promisify } from 'util'
import { MARKDOWN_MESSAGE } from './constants'

export const getRawDiff = async (
  targetBranch: string,
  filesPattern: string[]
) => {
  const filesPatternString = filesPattern.map((file) => `'${file}'`).join(' ')
  const { stdout, stderr } = await promisify(exec)(
    `git diff -w ${targetBranch} -- ${filesPatternString}`
  )
  if (stderr) {
    throw new Error(stderr)
  }
  return stdout
}

export const getMarkdownDiff = (rawDiff: string) => {
  const fileDiffs = rawDiff.split('diff --git a/')
  let markdownDiff = ''
  const hasDetectedChanges = rawDiff.length > 0
  for (const fileDiff of fileDiffs.slice(1)) {
    // skip the first element as it's empty
    const lines = fileDiff.split('\n')
    // get the file name from the first line
    const fileName = lines[0].split(' ')[0] 
    // filter the diff lines to only include the ones that start with + or -
    const filteredLines = lines
      .slice(1)
      .filter((line) => /^[\+-]{1}[^+-]/.test(line)) ?? MARKDOWN_MESSAGE.CHANGES_FALLBACK
    markdownDiff += `#### \`${fileName}\`\n\`\`\`diff\n${filteredLines.join(
      '\n'
    )}\n\`\`\`\n`
  }
  return hasDetectedChanges
    ? `${MARKDOWN_MESSAGE.CHANGES_DETECTED}\n${markdownDiff}`
    : MARKDOWN_MESSAGE.NO_CHANGES
}
