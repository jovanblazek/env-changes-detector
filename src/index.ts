import { MARKDOWN_MESSAGE, OUTPUT } from './constants'
import { getInput, setFailed, setOutput } from '@actions/core'
import { exec } from 'child_process'
import { promisify } from 'util'

export const run = async () => {
  try {
    const targetBranch = getInput('target-branch')
    const filesToCheck: string[] = JSON.parse(getInput('files'))
    setOutput(OUTPUT.HAS_DETECTED_CHANGES, false)
    setOutput(OUTPUT.RAW, [])
    setOutput(OUTPUT.MARKDOWN, MARKDOWN_MESSAGE.NO_CHANGES)

    const diffResult = await promisify(exec)(
      `git diff -w origin/${targetBranch} -- ${filesToCheck
        .map((file) => `'${file}'`)
        .join(' ')}`
    )
    if (diffResult.stderr) {
      throw new Error(diffResult.stderr)
    }

    if (diffResult.stdout === '') {
      return
    }

    const regex = /^(?<diff>[+-]{1}\w.*)|(?:diff --git\sa(?<file>.*?)\s.*)$/gm
    const matches = Array.from(
      diffResult.stdout.matchAll(regex),
      (match) => match.groups as { [key: string]: string }
    )

    // format found changes to markdown syntax
    const markdownMessage = matches.map((match, index) => {
      if (match.file) {
        if (index === 0) {
          return `\`${match.file}\``
        }
        return `\`\`\`\n\n\`${match.file}\``
      }
      const previousMatch = matches[index - 1]
      if (previousMatch.diff) {
        return match.diff
      }
      return `\`\`\` diff\n${match.diff}`
    })
    markdownMessage.push(`\`\`\``) // close last code block

    setOutput(OUTPUT.HAS_DETECTED_CHANGES, true)
    setOutput(OUTPUT.RAW, matches)
    setOutput(
      OUTPUT.MARKDOWN,
      `${MARKDOWN_MESSAGE.CHANGES_DETECTED}\n\n${markdownMessage.join('\n')}`
    )
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {
    setFailed(error)
  }
}

run()
