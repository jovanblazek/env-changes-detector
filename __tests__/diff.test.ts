import { MARKDOWN_MESSAGE } from '../src/constants'
import { getMarkdownDiff, getRawDiff } from '../src/diff'
import { exec } from 'child_process'
import { getFixture } from './utils'
jest.mock('child_process')

describe('diff', () => {
  describe('getRawDiff', () => {
    it('should return the stdout when the command executes successfully', async () => {
      const mockExec = exec as jest.MockedFunction<typeof exec>
      mockExec.mockImplementation((_, callback) =>
        // @ts-expect-error mock implementation
        callback(null, { stdout: 'mock stdout', stderr: '' })
      )

      const result = await getRawDiff('main', ['*.md', '**.ts'])
      expect(result).toBe('mock stdout')
      expect(mockExec).toHaveBeenCalledWith(
        `git diff -w -M100% origin/main -- '*.md' '**.ts'`,
        expect.any(Function)
      )
    })

    it('should throw an error when the command fails', async () => {
      const mockExec = exec as jest.MockedFunction<typeof exec>
      mockExec.mockImplementation((_, callback) =>
        // @ts-expect-error mock implementation
        callback(null, { stdout: '', stderr: 'mock error' })
      )

      await expect(getRawDiff('main', ['*.md', '**.ts'])).rejects.toThrow(
        'mock error'
      )
      expect(mockExec).toHaveBeenCalledWith(
        `git diff -w -M100% origin/main -- '*.md' '**.ts'`,
        expect.any(Function)
      )
    })
  })

  describe('getMarkdownDiff', () => {
    it('should return "No changes detected" when there are no changes', () => {
      const rawDiff = ''
      const result = getMarkdownDiff(rawDiff)
      expect(result).toBe(MARKDOWN_MESSAGE.NO_CHANGES)
    })

    it.each([
      ['file is added', 'new-file'],
      ['file is deleted', 'deleted-file'],
      ['file is renamed', 'renamed-file'],
      ['there are multiple changed lines in one file', 'changed-lines'],
      ['there are multiple changed files', 'all']
    ])('should return a markdown diff when the %s', (_, fixtureName) => {
      const rawDiff = getFixture(fixtureName)
      const result = getMarkdownDiff(rawDiff)
      expect(result).toMatchSnapshot()
    })

    it('should not include unchanged lines in the markdown diff', () => {
      const rawDiff = getFixture('unchanged-line')
      const result = getMarkdownDiff(rawDiff)
      console.log('result\n', result)
      expect(result).toMatchSnapshot()
    })
  })
})
