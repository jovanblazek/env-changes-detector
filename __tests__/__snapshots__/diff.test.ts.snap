// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff getMarkdownDiff should not include unchanged lines in the markdown diff 1`] = `
"## Detected changes in env files:
#### \`unchangedLine.js\`
\`\`\`diff
+   console.log("hello world");
\`\`\`

"
`;

exports[`diff getMarkdownDiff should return a markdown diff when the file is added 1`] = `
"## Detected changes in env files:
#### \`newfile.md\`
\`\`\`diff
+ newfile
\`\`\`

"
`;

exports[`diff getMarkdownDiff should return a markdown diff when the file is deleted 1`] = `
"## Detected changes in env files:
#### \`deleteFile.md\`
\`\`\`diff
File was deleted.
\`\`\`

"
`;

exports[`diff getMarkdownDiff should return a markdown diff when the file is renamed 1`] = `
"## Detected changes in env files:
#### \`newfile.md -> rename.md\`
\`\`\`diff
File was renamed.
\`\`\`

"
`;

exports[`diff getMarkdownDiff should return a markdown diff when the there are multiple changed files 1`] = `
"## Detected changes in env files:
#### \`deleteLine.md\`
\`\`\`diff
- newline
\`\`\`

#### \`deleteFile.md\`
\`\`\`diff
File was deleted.
\`\`\`

#### \`newfile.md\`
\`\`\`diff
+ newfile
\`\`\`

#### \`newLine.md\`
\`\`\`diff
+ newline
\`\`\`

#### \`newfile.md -> rename.md\`
\`\`\`diff
File was renamed.
\`\`\`

#### \`changedLines.md\`
\`\`\`diff
- oldLine
- oldLine2
+ changedLine
\`\`\`

"
`;

exports[`diff getMarkdownDiff should return a markdown diff when the there are multiple changed lines in one file 1`] = `
"## Detected changes in env files:
#### \`changedLines.md\`
\`\`\`diff
- oldLine
- oldLine2
+ changedLine
\`\`\`

"
`;
